miner/agent.go

```diff
@@ -58,8 +61,12 @@ func (self *CpuAgent) Stop() {
 }
 
 func (self *CpuAgent) Start() {
-	self.mu.Lock()
 	defer self.mu.Unlock()
+	self.mu.Lock()
+
+	if atomic.LoadInt32(&self.isMining) == 1 {
+		return // agent already started
+	}
```

Comment:
I.e. Either protect using the mutex only, and obtain the lock at the other side too, or if you use atomic operations that make sure they stand their place by themselves :)